### PR TITLE
[IoTeX ] support IoTeX native staking

### PR DESCRIPTION
**Pull request by lzxm160**
## Description
Add support IoTeX native staking

## Testing instructions
Unit tests in tests/IoTeX

## Types of changes
Refer to src/proto/IoTeX.proto

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description (e.g. 'Fixes #444 ).

### CODE REVIEW

**Comment by lzxm160** [comment]
pls wait me to add ios tests

File: src/IoTeX/Staking.cpp [hunk 1]
```diff
@@ -5,56 +5,120 @@
...
-    func.addInParam(std::make_shared<ParamByteArrayFix>(12, candidate));
-    func.addInParam(std::make_shared<ParamUInt256>(uint256_t(stakeDuration)));
-    func.addInParam(std::make_shared<ParamBool>(nonDecay));
-    func.addInParam(std::make_shared<ParamByteArray>(dataIn));
+const char* charFromTWData(TWData* _Nonnull data) {
+    return TWStringUTF8Bytes(TWStringCreateWithRawBytes(TWDataBytes(data), TWDataSize(data)));
+}
 
-    func.encode(dataOut);
+TWData* _Nullable stakingCreate(TWData* _Nonnull candidate, TWData* _Nonnull amount,
```
Comment (hewigovens) [inline src/IoTeX/Staking.cpp:19]: `TWData` is C struct, you should use `Data` here

File: src/proto/IoTeX.proto [hunk 2]
```diff
@@ -9,45 +9,68 @@ message Transfer {
...
+message Execution {
+  string amount = 1;
+  string contract = 2;
+  bytes data = 3;
+}
 
-    string amount  = 1;
-    string contract = 2;
+// create stake
+message StakeCreate {
```
Comment (hewigovens) [inline src/proto/IoTeX.proto:19]: can you nest them under `Staking`?

File: src/IoTeX/Staking.cpp (hunk 1 above)
Comment (lzxm160) [reply src/IoTeX/Staking.cpp:19]: sure

File: src/proto/IoTeX.proto (hunk 2 above)
Comment (lzxm160) [reply src/proto/IoTeX.proto:19]: ok,change it later

**Comment by lzxm160** [comment]
I'll fix ios test tomorrow

File: src/proto/IoTeX.proto [hunk 3]
```diff
@@ -78,3 +108,30 @@ message SigningOutput {
     // Signed Action hash
     bytes hash = 2;
 }
```
Comment (dustinxie) [inline src/proto/IoTeX.proto:110]: remove Staking staking = 11; 
and add new message 40~48 like ActionCore below

File: src/IoTeX/Signer.cpp [hunk 4]
```diff
@@ -44,41 +44,85 @@ Data Signer::hash() const {
...
-        auto encoded = encodeStaking(staking);
-        auto& execution = *action.mutable_execution();
-        execution.set_amount(staking.amount());
-        execution.set_contract(staking.contract());
-        execution.set_data(encoded.data(), encoded.size());
-    } else {
-        // ActionCore is almost same as SigningInput, missing field privateKey = 5;
-        action.ParseFromString(input.SerializeAsString());
-        action.DiscardUnknownFields();
+    action.set_version(input.version());
```
Comment (dustinxie) [inline src/IoTeX/Signer.cpp:48]: with updated ActionCore and SigningInput (see comment below)
we don't need to assert the type of input for staking
can directly use the else {} part of old code

**Comment by lzxm160** [comment]
all done,pls help review @hewigovens

**Comment by hewigovens** [comment]
will take a look soon

File: src/proto/IoTeX.proto [hunk 5]
```diff
@@ -10,43 +10,73 @@ message Transfer {
...
+
+  // add the amount of bucket
+  message StakeAddDeposit {
+    uint64 bucketIndex = 1;
+    string amount = 2;
+    bytes payload = 3;
+  }
+
+  // restake the duration and autoStake flag of bucket
+  message StakeRestake {
```
Comment (hewigovens) [inline src/proto/IoTeX.proto:36]:
```suggestion
  message Restake {
```
same for other namings

File: tests/IoTeX/TWIoTeXStakingTests.cpp [hunk 6]
```diff
@@ -1,156 +0,0 @@
...
-
-#include <gtest/gtest.h>
-
-using namespace TW;
-using namespace TW::IoTeX;
-
-static const char *_Nonnull IOTEX_STAKING_CONTRACT = "io1xpq62aw85uqzrccg9y5hnryv8ld2nkpycc3gza";
-static const char * IOTEX_STAKING_TEST = "this is a test";
-
-TEST(TWIoTeXStaking, SignStake) {
```
Comment (hewigovens) [inline tests/IoTeX/TWIoTeXStakingTests.cpp:22]: let's keep one test using `TWAnySignerSign`

**Comment by hewigovens** [comment]
please also add Kotlin tests

File: src/IoTeX/Staking.cpp [hunk 7]
```diff
@@ -5,56 +5,108 @@
...
-    func.addInParam(std::make_shared<ParamBool>(nonDecay));
-    func.addInParam(std::make_shared<ParamByteArray>(dataIn));
+const Data& dataFromString(const std::string& d) {
+    Data* data = new Data;
+    std::copy(d.c_str(), d.c_str() + d.length(), back_inserter(*data));
+    return *data;
+}
 
-    func.encode(dataOut);
+const Data& stakingCreate(const Data& candidate, const Data& amount, uint32_t duration,
```
[the remark this answers is not in the document]
Comment (lzxm160) [reply src/IoTeX/Staking.cpp:27]: not very big,you mean just return Data?

File: tests/IoTeX/TWIoTeXStakingTests.cpp (hunk 6 above)
Comment (lzxm160) [reply tests/IoTeX/TWIoTeXStakingTests.cpp:22]: sure

File: src/IoTeX/Staking.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (hewigovens) [reply src/IoTeX/Staking.cpp:27]: Yes

File: swift/Tests/Blockchains/IoTeXTests.swift [hunk 8]
```diff
@@ -29,81 +28,182 @@ class IoTeXTests: XCTestCase {
...
+        input.stakeCreate = IoTeXStaking.StakeCreate.with {
+            $0.candidateName = "io19d0p3ah4g8ww9d7kcxfq87yxe7fnr8rpth5shj"
+            $0.stakedAmount = "100"
+            $0.stakedDuration = 10000
+            $0.autoStake = true
+            $0.payload = "payload".data(using: .utf8)!
+        }
+        let output: IoTeXSigningOutput = AnySigner.sign(input: input, coin: .ioTeX)
+
+        XCTAssertEqual(output.encoded.hexString, "0a4b080118c0843d22023130c2023e0a29696f313964307033616834673877773964376b63786671383779786537666e7238727074683573686a120331303018904e20012a077061796c6f6164124104755ce6d8903f6b3793bddb4ea5d3589d637de2d209ae0ea930815c82db564ee8cc448886f639e8a0c7e94e99a5c1335b583c0bc76ef30dd6a1038ed9da8daf331a412e8bac421bab88dcd99c26ac8ffbf27f11ee57a41e7d2537891bfed5aed8e2e026d46e55d1b856787bc1cd7c1216a6e2534c5b5d1097c3afe8e657aa27cbbb0801")
```
[the remark this answers is not in the document]
Comment (lzxm160) [reply swift/Tests/Blockchains/IoTeXTests.swift:48]: Can this be added in next pr, as these interfaces are currently being tested

[the remark this answers is not in the document]
Comment (hewigovens) [reply swift/Tests/Blockchains/IoTeXTests.swift:48]: ok

**Comment by lzxm160** [comment]
> please also add Kotlin tests

all done,pls check

**Comment by lzxm160** [comment]
@madcake @catenocrypt pls help review,thanks

**Comment by lzxm160** [comment]
@hewigovens I don't have write access right,can you help?

**Pull request merged**